platooning_control_ihp/src/platoon_control_ihp_worker.cpp

```diff
@@ -22,30 +22,98 @@
    
-	PlatoonControlIHPWorker::PlatoonControlIHPWorker()
+	PlatoonControlWorkerIHP::PlatoonControlWorkerIHP()
     {
         pid_ctrl_ = PIDController();
         pp_ = PurePursuit();
 
     }
 
-    void PlatoonControlIHPWorker::updateConfigParams(PlatoonControlIHPPluginConfig new_config)
+    void PlatoonControlWorkerIHP::updateConfigParams(PlatooningControlIHPPluginConfig new_config)
     {
         ctrl_config_ = new_config;
         pid_ctrl_.config_ = new_config;
         pp_.config_ = new_config;
     }
 
-	double PlatoonControlIHPWorker::getLastSpeedCommand() const {
+	double PlatoonControlWorkerIHP::getLastSpeedCommand() const {
         return speedCmd_;
     }
 
-    void PlatoonControlIHPWorker::setCurrentPose(const geometry_msgs::PoseStamped msg)
+    void PlatoonControlWorkerIHP::setCurrentPose(const geometry_msgs::PoseStamped msg)
 	{
 		current_pose_ = msg.pose;
 	}
 
-    void PlatoonControlIHPWorker::generateSpeed(const cav_msgs::TrajectoryPlanPoint& point)
+    // UCLA: Put the IHP gap regulation at low level control plug-in
+    double PlatoonControlWorkerIHP::getIHPDesPosFollower(double leaderCurrentPosition)
+    {
+        /**
+         * Calculate desired position based on previous vehicle's trajectory for followers.
+         * 
+         * TODO: The platoon trajectory regulation is derived with the assumption that all vehicle 
+         *       have identical length (i.e., 5m). Future development is needed to include variable 
+         *       vehicle length into consideration.
+         */
+
+
+        // 1. find the summation of "veh_len/veh_speed" for all predecessors
+        double tmp_time_hdw = current_predecessor_time_headway_sum_;
+
+        // 2. read host veh and front veh info 
+        // Predecessor vehicle data.
+        double pred_spd = predecessor_speed_;       // m/s 
+        double pred_pos = predecessor_position_;    // m
+        
+        // host data. 
+        double ego_spd = currentSpeed;                          // m/s
+        double ego_pos = leaderCurrentPosition - actual_gap_;   // m
+        
+        // platoon position index
+        int pos_idx = host_platoon_position_;
+
+        double desirePlatoonGap = ctrl_config_.intra_tau;       //s
+        
+        // IHP desired position calculation methods
+        double pos_g;      // desired downtrack position calculated with time-gap, in m.
+        double pos_h;      // desired downtrack position calculated with distance headaway, in m.
+
+        // 3. IHP gap regualtion 
+        // intermediate variables 
+        double timeGapAndStepRatio = desirePlatoonGap/ctrl_config_.time_step;     // The ratio between desired platoon time gap and the current time_step.
+        double totalTimeGap = desirePlatoonGap*pos_idx;                           // The overall time gap from host vehicle to the platoon leader, in s.
+
+        // calcilate pos_gap and pos_headway
+        if ((pred_spd <= ego_spd) && (ego_spd <= ctrl_config_.ss_theta))
```

Comment:
It is unclear how what this condition is checking. 
In addition, you are calculating the pose_g and pse_h which are dependent on the gap, but I don't see the speed being used anywhere, can you explain?

Reply:
Agree, there needs to be comments explaining what is going on here. The design doc doesn't mention anything about two different conditions for this calc.